### ISSUE TITLE
Minifiers [ideas]

### ISSUE THREAD

**Issue by maltsev** [maltsev/htmlnano#12]
- [ngx_pagespeed](http://ngxpagespeed.com/)
- [html-minifier](https://www.npmjs.com/package/html-minifier)

**Comment by voischev** [comment]
remove \n \t and ect
```js
tree.match(/^\n\s*$/, () => '');
```
https://github.com/posthtml/project-stub/blob/master/gulpfile.js#L14-L18

**Comment by maltsev** [comment]
@voischev the module `collapseWhitespace` does that. Or did you find some error?

**Comment by stevenvachon** [comment]
The biggest thing for me is to split out to separate plugins for individual test suites:
* htmlnano-remove-empty-attrs
* htmlnano-normalize-whitespace
* htmlnano-merge-styles
* htmlnano-merge-scripts
* htmlnano-cssnano
* htmlnano-uglifyjs
* etc

@maltsev [normalize-html-whitespace](https://npmjs.com/normalize-html-whitespace) is the most accurate. Also, the key is to *normalize*, not *strip* as to preserve rendered appearance.

**Comment by maltsev** [comment]
@stevenvachon, first of all, thank you for new plugin ideas! I'll consider splitting out the modules to separate plugins. But it'll be much harder for me to maintain a few dozens small plugins (documentation, infrastructure) instead of one `htmlnano`.

**Comment by stevenvachon** [comment]
No problem. It's definitely more work to create separate projects, at least initially. Once created, it will make dealing with tests much easier because there're fewer of them and they're all very specific to one feature. This is how cssnano and cssnext are designed.

**Comment by maltsev** [comment]
@stevenvachon, I agree with you. I'm also trying to write the modules and the tests as separated as possible, so it'll be easy to split off them later.

**Comment by stevenvachon** [comment]
Whitespace within attributes containing space-separated lists can also be normalized:
```html
<a href="#" class="class1   class2 " rel=" external    nofollow ">link</a>
```
->
```html
<a href="#" class="class1 class2" rel="external nofollow">link</a>
```

**Comment by stevenvachon** [comment]
1. remove quotes on qualifying attributes
2. switch between `'` and `"` (per attribute) as attribute value encapsulators to find which produces the smallest output (due to any necessary commenting: `attr='blah \'blah\''` vs `attr="blah 'blah'"`)

**Comment by maltsev** [comment]
@stevenvachon, thanks! I've created an issue for your first idea: #27
An issue for the second one is already there: #6

**Comment by stevenvachon** [comment]
[svgo](https://npmjs.com/svgo) for inline svg

**Comment by stevenvachon** [comment]
1. Remove optional closing tags such as `</html>`, `</head>`, `</body>`, `</option>`, `</p>`, `</thead>`, `</tbody>`, `</tfoot>`, `</tr>`
2. remove optional opening tags such as `<html>`, `<head>`, `<body>`

**Comment by stevenvachon** [comment]
Strip HTML from content within `<title></title>` tags as browsers and search engines will ignore it anyway.
